libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -881,6 +915,75 @@ string YulUtilFunctions::readFromStorageDynamic(Type const& _type, bool _splitFu
 	});
 }
 
+string YulUtilFunctions::readFromMemory(Type const& _type)
+{
+	return readFromMemoryOrCalldata(_type, false);
+}
+
+string YulUtilFunctions::readFromCalldata(Type const& _type)
+{
+	return readFromMemoryOrCalldata(_type, true);
+}
+
+string YulUtilFunctions::readFromMemoryOrCalldata(Type const& _type, bool _fromCalldata)
+{
+	string functionName =
+		string("read_from_") +
+		(_fromCalldata ? "calldata" : "memory") +
+		"_dynamic_" +
+		_type.identifier();
+
+	return m_functionCollector->createFunction(functionName, [&] {
+		if (auto arrayType = dynamic_cast<ArrayType const*>(&_type))
+		{
+			solAssert(!arrayType->isDynamicallySized(), "");
+			solAssert(!_fromCalldata, "");
+			//solAssert(_padToWords, "");
+
+			return Whiskers(R"(
+				function <functionName>(addr) -> value {
+					value := mload(addr)
+				}
+			)")
+			("functionName", functionName)
+			.render();
+		}
+
+		unsigned numBytes = _type.calldataEncodedSize(false);
+
+		solAssert(numBytes != 0, "");
+
+		if (auto const* funType = dynamic_cast<FunctionType const*>(&_type))
+			if (funType->kind() == FunctionType::Kind::External)
+				return Whiskers(R"(
+					function <functionName>(addr) -> addr, selector {
+						addr := <load>(addr)
```

Comment:
Because of situations like that, I would propose to use `memptr` or `ptr` for "memory addresses" and `addr` only for account addresses.

Reply:
yeah, makes sense